### PR TITLE
Document Maxwell 2's texture headers

### PR DESCRIPTION
**Pull request by Gnurou**
This PR does two things:

1) Update the G80 header documentation to complete them and harmonize the names with NVIDIA's
2) Add the complete documentation for the GM2 texture headers. This new format is the only one recognized by GM200 and later chips.

The renaming of G80 is helpful in that it helps understanding where fields have moved in the new GM2 structures, however it will impact code using the generated headers. If I went too far in the renaming frenzy, please let me know.

Also I am not too sure whether my usage of the inline property in the new enums is correct.

### CODE REVIEW

**Comment by mupuf** [comment]
Thanks!

File: rnndb/g80_defs.xml [hunk 1]
```diff
@@ -72,17 +72,122 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
...
+	<value value="6" name="ONE_INT"/> <!-- 0x00000001 -->
+	<value value="7" name="ONE_FLOAT"/> <!-- 0x3f800000 -->
+</enum>
+
+<enum name="G80_TIC_TYPE">
+	<value value="1" name="SNORM"/>
+	<value value="2" name="UNORM"/>
+	<value value="3" name="SINT"/>
+	<value value="4" name="UINT"/>
+	<value value="5" name="SNORM_FORCE_FP16"/>
```
Comment (imirkin) [inline rnndb/g80_defs.xml]: Can you explain what this name means? Before we had SSCALED, but pretty sure we never used it, so it could well have been wrong.

File: rnndb/graph/g80_texture.xml [hunk 2]
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
...
-			<value value="0x32" name="Z24_X8_S8_C8_X16_MS8_CS8"/>
-			<value value="0x33" name="Z32_X8_C8_X16_MS4_CS4" />
-			<value value="0x34" name="Z32_X8_C8_X16_MS8_CS8" />
-			<value value="0x35" name="Z32_S8_C8_X16_MS4_CS4" />
-			<value value="0x36" name="Z32_S8_C8_X16_MS8_CS8" />
-			<value value="0x37" name="Z24_X8_S8_C8_X16_MS4_CS12"/>
-			<value value="0x38" name="Z32_X8_C8_X16_MS4_CS12" />
-			<value value="0x39" name="Z32_S8_C8_X16_MS4_CS12" />
-			<value value="0x3a" name="Z16" />
+		<bitfield pos="30" name="PACK_COMPONENTS">
```
Comment (imirkin) [inline rnndb/graph/g80_texture.xml]: Please confirm variants, unless you're saying this was available on G80

File: rnndb/graph/g80_texture.xml [hunk 3]
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
...
+			<value value="0x10" name="BC6H_SF16" variants="GF100-">
+				<doc>similar to ZOH (FP16)</doc>
+			</value>
+			<value value="0x11" name="BC6H_UF16" variants="GF100-">
+				<doc>similar to ZOH (UFP16)</doc>
+			</value>
+			<value value="0x17" name="BC7U" variants="GF100-">
+				<doc>similar to ZOL (UNORM8)</doc>
+			</value>
+			<value value="0x06" name="ETC2_RGB"/>
```
Comment (imirkin) [inline rnndb/graph/g80_texture.xml]: variants on all the ETC/EAC formats? guessing GK104+ or GM107+?

File: rnndb/graph/g80_texture.xml [hunk 4]
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
...
+			<value value="0x37" name="X8Z24_X16V8S8__COV4R12V">
+				<doc>stencil,    4+12 VCAA, 24-bit fixed-point depth</doc>
+			</value>
+			<value value="0x38" name="ZF32_X16V8X8__COV4R12V">
+				<doc>no stencil, 4+12 VCAA, 32-bit floating-point depth</doc>
+			</value>
+			<value value="0x39" name="ZF32_X16V8S8__COV4R12V">
+				<doc>stencil,    4+12 VCAA, 32-bit floating-point depth</doc>
+			</value>
+			<value value="0x3a" name="Z16">
```
Comment (imirkin) [inline rnndb/graph/g80_texture.xml]: variants (G200+? Similar for some of the coverage ones, pretty sure they're not all available on G80)

File: rnndb/graph/g80_texture.xml [hunk 5]
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
...
 		</bitfield>
 	</reg32>
 
 	<reg32 offset="0x10" name="4">
-		<bitfield high="31" low="0" name="WIDTH" />
+		<bitfield high="29" low="0" name="WIDTH" />
+		<bitfield pos="30" name="DEPTH_TEXTURE" type="boolean">
+			<doc>Tells texture unit to smear border color from R to RGBA channels</doc>
+		</bitfield>
+		<bitfield pos="31" name="USE_TEXTURE_HEADER_V2" type="boolean">
```
Comment (imirkin) [inline rnndb/graph/g80_texture.xml]: G84+?

File: rnndb/graph/g80_texture.xml [hunk 6]
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
...
+			<doc>spread modifier modifies fine spread only, not sample count</doc>
+		</bitfield>
+	</reg32>
+
 	<reg32 offset="0x1c" name="7">
-		<bitfield high="3" low="0" name="BASE_LEVEL" />
-		<bitfield high="7" low="4" name="MAX_LEVEL" />
-		<bitfield high="15" low="12" name="MS_MODE" type="g80_msaa_mode">
-			<doc>Like g80/gf100_msaa_mode, dimensions are unaffected.</doc>
+		<bitfield high="3" low="0" name="RES_VIEW_MIN_MIP_LEVEL">
```
Comment (imirkin) [inline rnndb/graph/g80_texture.xml]: I guess these should all be G84+, with G80 getting the color key thing? Technically it's not based on chipset, but in practice... Or you can put the variant on the whole reg and then make a G80-only reg that's just called COLOR_KEY or whatever.

File: rnndb/graph/g80_texture.xml [hunk 7]
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
...
+		<bitfield high="7" low="4" name="RES_VIEW_MAX_MIP_LEVEL">
+			<doc>max LOD for texture resource views</doc>
+		</bitfield>
+		<bitfield pos="8" name="HEIGHT_MSB">
+			<doc>bit 16 of the texture's height, bits 0-15 are in Dword5</doc>
+		</bitfield>
+		<bitfield high="15" low="12" name="MULTI_SAMPLE_COUNT" type="g80_msaa_mode">
+			<doc>Like g80/gf100_msaa_mode, dimensions are unaffected. return value for the TEX_HEADER_TEXTURE_TYPE ISA instruction</doc>
+		</bitfield>
+		<bitfield high="27" low="16" name="MIN_LOD_CLAMP">
```
Comment (imirkin) [inline rnndb/graph/g80_texture.xml]: type="ufixed" radix="8"

Please do this in other places where you have fixed point bitfields (use type="fixed" for signed)

File: rnndb/graph/gm200_texture.xml [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+				<doc>stencil,    8+24 VCAA, 24-bit fixed-point depth</doc>
+			</value>
+			<value value="0x3d" name="ZF32_X16V8X8__COV8R24V">
+				<doc>no stencil, 8+24 VCAA, 32-bit floating-point depth</doc>
+			</value>
+			<value value="0x3e" name="ZF32_X16V8S8__COV8R24V">
+				<doc>stencil,    8+24 VCAA, 32-bit floating-point depth</doc>
+			</value>
+
+			<value value="0x40" name="ASTC_2D_4X4"/>
```
Comment (imirkin) [inline rnndb/graph/gm200_texture.xml]: Please confirm that these are all available on GM107

File: rnndb/graph/gm200_texture.xml [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+			<doc>least significant 16 bits of the "width minus one" field</doc>
+		</bitfield>
+		<bitfield high="21" low="19" name="ANISO_SPREAD_MAX_LOG2">
+			<doc>Log(aniso-filter spread)</doc>
+		</bitfield>
+		<bitfield high="31" low="29" name="BORDER_SIZE">
+			<value value="0" name="ONE"/>
+			<value value="1" name="TWO"/>
+			<value value="2" name="FOUR"/>
+			<value value="3" name="HEIGHT"/>
```
Comment (imirkin) [inline rnndb/graph/gm200_texture.xml]: Maybe that's how it's spelled in french...

File: rnndb/graph/gm200_texture.xml [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+			<value value="7" name="SAMPLER_COLOR">
+				<doc>border comes from sampler</doc>
+			</value>
+			<doc>border may be either constant color (from sampler), 1, 2, 4, or 8</doc>
```
Comment (imirkin) [inline rnndb/graph/gm200_texture.xml]: If you're able to, please elaborate, as this comment makes no sense to me.

File: rnndb/graph/g80_texture.xml [hunk 11]
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
 
 <import file="nvchipsets.xml"/>
 <import file="g80_defs.xml"/>
+<import file="nv_defs.xml"/>
```
Comment (imirkin) [inline rnndb/graph/g80_texture.xml]: why is this necessary?

File: rnndb/graph/g80_texture.xml [hunk 12]
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
...
 	</reg32>
 
 	<reg32 offset="0x10" name="4">
-		<bitfield high="31" low="0" name="WIDTH" />
+		<bitfield high="29" low="0" name="WIDTH" />
+		<bitfield pos="30" name="DEPTH_TEXTURE" type="boolean">
+			<doc>Tells texture unit to smear border color from R to RGBA channels</doc>
+		</bitfield>
+		<bitfield pos="31" name="USE_TEXTURE_HEADER_V2" type="boolean">
+			<doc>if 0, then byte 7 of the TIC is used to store the color key value. Should always be 1.</doc>
```
Comment (koriakin) [inline rnndb/graph/g80_texture.xml]: you mean word 7, right?

File: rnndb/graph/g80_texture.xml [hunk 13]
```diff
@@ -170,45 +325,53 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
...
+		<bitfield pos="9" name="CUBEMAP_INTERFACE_FILTERING" type="boolean" variants="GK104-"/>
+		<bitfield high="24" low="12" name="MIP_LOD_BIAS" type="fixed" radix="8">
 			<doc>Fixed-point signed 1.4.8</doc>
 		</bitfield>
-		<bitfield pos="25" name="FORCE_NONNORMALIZED_COORDS" type="boolean" variants="GK104-"/>
+		<bitfield pos="25" name="FLOAT_COORD_NORMALIZATION" type="gk104_float_coord_normalization" variants="GK104-"/>
+		<bitfield high="30" low="26" name="TRILIN_OPT">
+			<doc>slope for trilinear optimization, e2m3, controlled by USE_HEADER_OPT_CONTROL</doc>
+		</bitfield>
 		<value value="0x10000000" name="UNKN_ANISO_15" />
```
Comment (koriakin) [inline rnndb/graph/g80_texture.xml]: those 2 should be killed off, they refer to TRILIN_OPT bits

File: rnndb/graph/gm200_texture.xml [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+	<value value="0x0" name="ONE_D_BUFFER"/>
+	<value value="0x1" name="PITCH_COLORKEY"/>
+	<value value="0x2" name="PITCH"/>
+	<value value="0x3" name="BLOCKLINEAR"/>
+	<value value="0x4" name="BLOCKLINEAR_COLORKEY"/>
+</enum>
+
+<domain name="TIC" size="0x20" prefix="chipset" variants="GM107-">
+	<brief>Texture image control block (for Maxwell)</brief>
+	<doc><p>For Maxwell, the texture headers were substantially redesigned, and five versions are now used. The HEADER_VERSION field of Dword 2 controls which version is used. Many fields change depending on the version, so one has to be extra careful with that.</p><p>GM10X still understands the older format, and uses it by default. It can optionally use the new format if the SetSelectMaxwellTextureHeaders method (address 0xf10) is called with an argument of 1. GPUs from GM20X only understand the new format, and SetSelectMaxwellTextureHeaders is a no-op.</p></doc>
```
Comment (koriakin) [inline rnndb/graph/gm200_texture.xml]: please add some newlines here

File: rnndb/graph/gm200_texture.xml [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+	<reg32 offset="0x4" name="1_ONE_D_BUFFER" varset="TIC_HEADER_VERSION" variants="ONE_D_BUFFER">
+		<brief>used if (HEADER_VERSION == ONE_D_BUFFER)</brief>
+		<bitfield high="31" low="0" name="ADDRESS_BITS_31_TO_0">
+			<doc>32 LSBs of the texture data, must specify down to the byte</doc>
+		</bitfield>
+	</reg32>
+
+	<reg32 offset="0x4" name="1_PITCH" varset="TIC_HEADER_VERSION" variants="PITCH_COLORKEY PITCH">
+		<brief>used if (HEADER_VERSION == PITCH_COLORKEY || HEADER_VERSION == PITCH)</brief>
+		<bitfield high="31" low="5" name="ADDRESS_BITS_31_TO_5">
```
Comment (koriakin) [inline rnndb/graph/gm200_texture.xml]: add shr="5" so that values will  be auto-shifted by 5

File: rnndb/graph/gm200_texture.xml [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+			<doc>27 MSBs of the lower dword of the address of the texture data, pitch textures are 32B-aligned. NOTE: due to a HW bug, pitch surfaces accessed with SUST must be 256B aligned</doc>
+		</bitfield>
+	</reg32>
+
+	<reg32 offset="0x4" name="1_BLOCKLINEAR" varset="TIC_HEADER_VERSION" variants="BLOCKLINEAR BLOCKLINEAR_COLORKEY">
+		<brief>used if (HEADER_VERSION == BLOCKLINEAR || HEADER_VERSION == BLOCKLINEAR_COLORKEY)</brief>
+		<bitfield high="6" low="5" name="GOB_DEPTH_OFFSET">
+			<doc>offset from the base in the depth dimension, to account for sub-allocation of textures with 3D gobs</doc>
+		</bitfield>
+		<bitfield high="31" low="9" name="ADDRESS_BITS_31_TO_9">
```
Comment (koriakin) [inline rnndb/graph/gm200_texture.xml]: also shr="9"

File: rnndb/graph/gm200_texture.xml [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+
+<enum name="TIC_HEADER_VERSION" inline="yes">
+	<value value="0x0" name="ONE_D_BUFFER"/>
+	<value value="0x1" name="PITCH_COLORKEY"/>
+	<value value="0x2" name="PITCH"/>
+	<value value="0x3" name="BLOCKLINEAR"/>
+	<value value="0x4" name="BLOCKLINEAR_COLORKEY"/>
+</enum>
+
+<domain name="TIC" size="0x20" prefix="chipset" variants="GM107-">
```
Comment (imirkin) [inline rnndb/graph/gm200_texture.xml]: I wonder if this should be called TIC2. On GM107 both this and the other one will be available, and I want to be able to include this file in rnndb/root.xml

File: rnndb/graph/gm200_texture.xml [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+	<reg32 offset="0xc" name="3_ONE_D_BUFFER" varset="TIC_HEADER_VERSION" variants="ONE_D_BUFFER">
+		<brief>these fields are available if (HEADER_VERSION == ONE_D_BUFFER)</brief>
+		<bitfield high="15" low="0" name="WIDTH_MINUS_ONE_BITS_31_TO_16">
+			<doc>most significant 16 bits of the "width minus one" field</doc>
+		</bitfield>
+	</reg32>
+
+	<reg32 offset="0xc" name="3_PITCH" varset="TIC_HEADER_VERSION" variants="PITCH_COLORKEY PITCH">
+		<brief>these fields are available if (HEADER_VERSION == PITCH_COLORKEY || HEADER_VERSION == PITCH)</brief>
+		<bitfield high="15" low="0" name="PITCH_BITS_20_TO_5">
```
Comment (koriakin) [inline rnndb/graph/gm200_texture.xml]: shr="5"

File: rnndb/graph/gm200_texture.xml [hunk 19]
```diff
@@ -0,0 +1,390 @@
...
+			<doc>width of block in gobs, with only one choice - IGNORED BY TEX (ASSUMED ONE_GOB)</doc>
+		</bitfield>
+		<bitfield high="5" low="3" name="GOBS_PER_BLOCK_HEIGHT" type="g80_gobs_per_block">
+			<doc>height of block in gobs</doc>
+		</bitfield>
+		<bitfield high="8" low="6" name="GOBS_PER_BLOCK_DEPTH" type="g80_gobs_per_block">
+			<doc>depth of block in gobs</doc>
+		</bitfield>
+		<bitfield high="12" low="10" name="TILE_WIDTH_IN_GOBS" type="g80_gobs_per_block">
+			<doc>width of the sparse tile in gobs</doc>
```
Comment (koriakin) [inline rnndb/graph/gm200_texture.xml]: huh, sparse tile?

File: rnndb/graph/g80_texture.xml (hunk 3 above)
Comment (imirkin) [inline rnndb/graph/g80_texture.xml]: FWIW after a bit of futzing on GK208 I couldn't get this to work. Certainly an option here is that I messed it up, but I was also seeing this in dmesg:
```
[175799.299477] nouveau 0000:01:00.0: gr: TRAP ch 2 [003fbe1000 oes_compressed_[6736]]
[175799.299484] nouveau 0000:01:00.0: gr: GPC0/TPC0/TEX: 80000009
[175799.299488] nouveau 0000:01:00.0: gr: GPC0/TPC1/TEX: 80000009
```
Guessing it's GM107+?

File: rnndb/g80_defs.xml (hunk 1 above)
Comment (Gnurou) [inline rnndb/g80_defs.xml]: The doc says "native 8/16bit snorm components forced to FP16 format". I have added that information, although the name already somehow contains it.

File: rnndb/graph/g80_texture.xml (hunk 2 above)
Comment (Gnurou) [inline rnndb/graph/g80_texture.xml]: This field seems to have appeared in G84.

File: rnndb/graph/g80_texture.xml (hunk 3 above)
Comment (Gnurou) [inline rnndb/graph/g80_texture.xml]: After some research it seems to have appeared with GK104+. Not sure why you would get this error. :/

File: rnndb/graph/g80_texture.xml (hunk 4 above)
Comment (Gnurou) [inline rnndb/graph/g80_texture.xml]: Not clear - but I cannot find any trace of these before fermi, so let's go with GF100+

File: rnndb/graph/g80_texture.xml (hunk 12 above)
Comment (Gnurou) [inline rnndb/graph/g80_texture.xml]: Correct, thanks.

File: rnndb/graph/g80_texture.xml (hunk 5 above)
Comment (Gnurou) [inline rnndb/graph/g80_texture.xml]: Seems so, yes.

File: rnndb/graph/g80_texture.xml (hunk 6 above)
Comment (Gnurou) [inline rnndb/graph/g80_texture.xml]: I made two variants of this register depending on the header format version used.

File: rnndb/graph/g80_texture.xml (hunk 7 above)
Comment (Gnurou) [inline rnndb/graph/g80_texture.xml]: Should be fixed throughout the patch.

File: rnndb/graph/g80_texture.xml (hunk 13 above)
Comment (Gnurou) [inline rnndb/graph/g80_texture.xml]: Done, thanks.

File: rnndb/graph/gm200_texture.xml (hunk 17 above)
Comment (Gnurou) [inline rnndb/graph/gm200_texture.xml]: You're right. Updated.

File: rnndb/graph/gm200_texture.xml (hunk 14 above)
Comment (Gnurou) [inline rnndb/graph/gm200_texture.xml]: Fixed formatting of other long lines as well.

File: rnndb/graph/gm200_texture.xml (hunk 8 above)
Comment (Gnurou) [inline rnndb/graph/gm200_texture.xml]: They are. Interestingly they also seem to be available on GK20A but not other Kepler chips.

File: rnndb/graph/gm200_texture.xml (hunk 15 above)
Comment (Gnurou) [inline rnndb/graph/gm200_texture.xml]: Done.

File: rnndb/graph/gm200_texture.xml (hunk 16 above)
Comment (Gnurou) [inline rnndb/graph/gm200_texture.xml]: Done.

File: rnndb/graph/gm200_texture.xml (hunk 18 above)
Comment (Gnurou) [inline rnndb/graph/gm200_texture.xml]: Fixed.

File: rnndb/graph/gm200_texture.xml (hunk 9 above)
Comment (Gnurou) [inline rnndb/graph/gm200_texture.xml]: Urgh, fixed. :)

File: rnndb/graph/gm200_texture.xml (hunk 10 above)
Comment (Gnurou) [inline rnndb/graph/gm200_texture.xml]: Cannot find more information about this, but your explanation is the only one that makes sense.

File: rnndb/graph/g80_texture.xml (hunk 11 above)
Comment (Gnurou) [inline rnndb/graph/g80_texture.xml]: rnn will complain about inv_memory_layout being an unknown type without this. But this was already the case before this change, so let me move this into its own patch.

**Comment by Gnurou** [comment]
Pushed v2 addressing all the comments received. Also added a missing field in g80 that only exists on GK20A (FWIW) and merged all register variants for gm200 as suggested by mwk on IRC.
